man/debianize.1
.TH debianize 1 "2004-09-10" "logilab devtools"

.SH NAME
debianize \- automatic maintenance of a package's debian files


.SH SYNOPSIS
.B debianize
.RI [ options ]
.br

.SH DESCRIPTION
Part of logilab's devtools.

.SH LICENCE
(c) 2004 Logilab. General Public Licence.

.SH SEE ALSO
/usr/share/doc/pythonX.Y-devtools/

.SH AUTHOR
Logilab <contact@logilab.fr>


ChangeLog
ChangeLog for logilab.devtools
==============================

--
    * debianize: fix the way to handle subpackage's __init__.py file (there 
      is now a subpackage_master boolean property in the __pkginfo__.py which
      tells if a package is handling the __init__.py file, so now only one 
      subpackage should set this to True and the others should depends on 
      this package)
    * debianize: fix Uploader to Uploaders in control
	* pkginfo: fix debian handler detection
    * makedistrib: don't ask to tag package if an error occurs
	
2006-01-10  --  0.8.0
    * debianize:
      * updated to handle site-python installation with architecture
        independant package (implements #10119)
      * consider the optional debian_uploader variable in package
        information 
    * fix .examples files generation with debianize
    * fix templates: control.standard-version is now 3.6.2, updated FSF
        address in debian licenses files
    * fix package type detection in preparedistrib (fix #9917)
	

2005-07-29  --  0.7.0
    * fix dh -i/-a option handing in rules file generated by debianize
      (#9700)
    * no more empty lines annoying emacs in rules files generated by
      debianize  (#9710)
    * better use of .docs files / dh_installdocs and .examples files
      / dh_installexamples in debianize
    * better pbuilder support in buildeb (#1045)
    * tagpackage try to tag only distributed files (#674)
    * catch EOFError and KeyboardInterrupt when waiting for an input to
      avoid traceback
    * fix svn_agent.log_info message format
	
2005-07-19  --  0.6.7
    * fix call to pycoverage with missing arguments (close #9584)
	
    * makedistrib force manifest regeneration when building debian
      packages (close #9134)


	
2005-06-20  --  0.6.6
    * fix for apycot checkers (python2.2 compat + handle apycot >= 0.7)

	

2005-06-14  --  0.6.5
    * new options on checkpackage to select tests to apply
	
    * handle optional dependancy on logilab-common for the -test package

    * fix svn tagging (both svn and cvs tags now include the
      project's name)

    * fix SGML catalog checking for registered DTDs with another language
      than EN



2005-02-25  --  0.6.4
    * fixed various bugs in lib.changelog
    * added write-permission check before checkin


	
2005-02-16  --  0.6.3
    * fix pycoverage to not include python file which are not in a
      subpackage (i.e. directory without __init__.py  file)

    * debianize: fix html doc detection, fix bug with py2.3 and optik

    * fix emacs startup file so that cover mode is automatically loaded at
      startup


	
2005-01-21  --  0.6.2
    * fix svn tag command

    * fix cvs edited command

    * added copyright information for coverage.py in the README and in
      the debian copyright files
	
    * add 2.4 generation by default to debianize

    * fix ChangeLog and debian/changelog parsers/writers to be less
      intrusive
	  
    * new column %missing in pycoverage reports

    * included a new emacs mode for coverage files

	
	  
2004-12-22  --  0.6.1
    * add missing import to vcslib.cvs (fix #8520)



2004-12-16  --  0.6.0
    * using optparse instead of getopt in dos2unix, added "replacement"
      and "listing" options

    * added the missing dos2unix script to the distribution

    * allow comments in DEPENDS, RECOMMENDS and SUGGEST files (using #
      at the begining of the comment line)

    * check_manifest and check_bin ignore .svn directory

    * removed the_coverage global in lib/coverage.py

    * fix max length and bullet lists in debian's descriptions

    * added missing end line in default debian/watch

    * major debianize refactoring using class / inheritance instead of
      functions composition

    * import the version control library from oobrother extended with code
      from devtools / apycot

    * support to read/write debian changelog file (see the --debian option
      of the changelog command)

    * consider lines beginning with a # as comment in DEPENDS, RECOMMENDS
      and SUGGESTS files



2004-11-03  --  0.5.4
    * fix __file__ problem with pycoverage

    * fixed -h option of pycoverage

    * checkpackage display output of executable which don't return 0 status
      when executed using -h or -help

    * tool 'dos2unix' added: removes '\r' inside specified files
      or all files of specified folders (which may be browsed
      recursively with option -r)



2004-10-18  --  0.5.3
    * licenses are now defined as text files in templates/licenses, so
      providing a way to recognize some new licenses

    * make buildeb, debianize and pycoverage more easily usable as a
      library

    * minor fix in changelog parsing

    * use new format in debian/watch

    * fix debianize's man pages handling



2004-07-07  --  0.5.2
    * fix cvs log parsing

    * fixed buildpackage with sdist target

    * buildeb may take a "command" argument, and use os.system instead of
      spawnv

    * pycoverage ignore "thirdparty" directory

    * spell check descriptions if snakespell is installed



2004-02-19  --  0.5.1
    * fix and enhance announce generation

    * fix default documentation and html_documentation callbacks for pkginfo
      to avoid error when launched from outside of the package directory

    * handle recommands and suggests on the dummy package



2004-02-17  --  0.5.0
    * added a new variable in __pkginfo__, distname. It allows to specify
      a distribution name different than the package name (for instance,
      apycot/tester). If not set, this variable default to the same as
      modname.

    * ignore ,cover files in manifest checking

    * pycoverage sort result by % coverage instead of module names

    * fixed debianize to get package's homepage at the last line of the long
      description



2004-02-13  --  0.4.5
    * fix debianize to handle dependancies to python standalone package

debianize.py
#!/usr/bin/env python
# Copyright (c) 2003-2004 LOGILAB S.A. (Paris, FRANCE).
# http://www.logilab.fr/ -- mailto:contact@logilab.fr
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 2 of the License, or (at your option) any later
# version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
"""Debianize a Python / Zope package

USAGE: debianize [OPTIONS] [package_dir]
          create / update debian files (debian mode). If no python versions are
          provided, default to 2.1 2.2 2.3 (on the first run). Used options are
          saved in a config file (debian/debianizerc) so you don't have to
          provide them on futur runs.
          The <python version> arguments is only used with the python-library
          handler.  
          
       debianize [OPTIONS] [package_dir] prepare 
          create a directory to build the package (prepare mode).

Both commands should be run in the root of the source package.

OPTIONS:
  -h / --help
       display this help message and exit.

  -o / --orig
       create a .orig.tar.gz (only in prepare mode).

EXAMPLE:
$ cd logilab/pyreverse
$ debianize 2.2 2.3
$ cd `debianize prepare --orig`
$ fakeroot make -f debian/rules binary
"""

__revision__ = '$Id: debianize.py,v 1.30 2004-11-10 12:02:12 syt Exp $'

import sys
import os
import getopt
from os.path import join

from logilab.devtools.lib import TextReporter
from logilab.devtools.lib.pkginfo import PackageInfo
from logilab.devtools.debhdlrs import get_package_handler, replace, empty

REPORTER = TextReporter()
DEBIAN_DIR = 'debian'
    
def debianize(command='update', reporter=REPORTER,
              directory=None, orig=False,
              replace_func=replace, empty_func=empty):
    directory = directory or os.getcwd()
    pkginfo = PackageInfo(reporter, directory)
    pkghandler = get_package_handler(pkginfo, replace_func, empty_func)
    debian_dir = join(directory, DEBIAN_DIR)
    if command == 'update':
        pkghandler.update_debian_files(debian_dir)
    else: # command == 'prepare'
        print pkghandler.prepare_debian(orig)


def run(args):
    """debianize a package according to command line arguments"""
    orig = False
    directory = None
    options, args = getopt.getopt(args, 'ho', ['help', 'orig'])
    for option in options:
        if option[0] in ('-h', '--help'):
            print __doc__
            return
        if option[0] in ('-o', '--orig'):
            orig = True
    command = 'update'
    try:
        args.remove('update')
    except ValueError:
        pass
    if args:
        try:
            args.remove('prepare')
            command = 'prepare'
        except ValueError:
            command = 'update'
        if len(args) > 1:
            raise Exception('Too many arguments %s' % ' '.join(args[1:]))
        elif args:
            directory = args[0]
    print '*' * 80
    print 'creating/updating debian files'
    debianize(command, directory=directory, orig=orig)
        
if __name__ == '__main__':
    run(sys.argv[1:])
